man/dmfeed.1
.TH "DMFEED" "1" "" "Version 2.0.0" "User Commands"
.SH NAME
dmfeed \(en creates Atom XML feed from log messages
.SH SYNOPSIS
\f[B]dmfeed\f[R] \-\-help
.PP
\f[B]dmfeed\f[R] \-\-version
.PP
\f[B]dmfeed\f[R] \-\-\f[B]database\f[R] \f[I]file\f[R]
[\-\-\f[B]output\f[R] \f[I]file\f[R]] [\-\-\f[B]node\f[R] \f[I]id\f[R]]
[\-\-\f[B]minlevel\f[R] \f[I]level\f[R]] [\-\-\f[B]maxlevel\f[R]
\f[I]level\f[R]] [\-\-\f[B]entries\f[R] \f[I]count\f[R]]
[\-\-\f[B]author\f[R] \f[I]name\f[R]] [\-\-\f[B]email\f[R]
\f[I]address\f[R]] [\-\-\f[B]id\f[R] \f[I]uuid\f[R]]
[\-\-\f[B]title\f[R] \f[I]title\f[R]] [\-\-\f[B]subtitle\f[R]
\f[I]title\f[R]] [\-\-\f[B]url\f[R] \f[I]url\f[R]] [\-\-\f[B]xsl\f[R]
\f[I]file\f[R]]
.PP
\f[B]dmfeed\f[R] \-\-\f[B]config\f[R] \f[I]file\f[R] [\-\-\f[B]name\f[R]
\f[I]name\f[R]]
.SH DESCRIPTION
This program creates a web feed from log messages in Atom Syndication
Format (RFC 4287).
The log messages are read from database and written as XML to standard
output or file.
.PP
The feed id has to be a 36 characters long UUID with hyphens.
News aggregators use the id to identify the feed.
Therefore, the id should not be reused among different feeds.
Run \f[I]dmuuid(1)\f[R] to generate a valid UUID.
.PP
The time stamp of the feed in the updated element is set to the date and
time of the last log message.
If no logs have been added to the database since the last file
modification of the feed, the output file is not updated, unless
argument \f[CR]\-\-force\f[R] is passed.
.SH OPTIONS
.TP
\f[B]\-\-author\f[R], \f[B]\-A\f[R] \f[I]name\f[R]
Name of feed author or organisation.
.TP
\f[B]\-\-config\f[R], \f[B]\-c\f[R] \f[I]file\f[R]
Path to configuration file.
.TP
\f[B]\-\-database\f[R], \f[B]\-d\f[R] \f[I]file\f[R]
Path to SQLite log database.
.TP
\f[B]\-\-email\f[R], \f[B]\-M\f[R] \f[I]address\f[R]
E\-mail address of feed author.
.TP
\f[B]\-\-entries\f[R], \f[B]\-E\f[R] \f[I]count\f[R]
Maximum number of entries in feed (default is 50).
.TP
\f[B]\-\-force\f[R], \f[B]\-F\f[R]
Force writing of output file.
If not set, the output file will be updated only if new log records are
available.
.TP
\f[B]\-\-help\f[R], \f[B]\-h\f[R]
Print available command\-line arguments and quit.
.TP
\f[B]\-\-id\f[R], \f[B]\-I\f[R] \f[I]uuid\f[R]
UUID of the feed, 36 characters long with hyphens.
.TP
\f[B]\-\-maxlevel\f[R], \f[B]\-K\f[R] \f[I]level\f[R]
Select log messages of the given maximum log level (from 1 to 5).
Must be greater or equal the minimum level.
The argument may be an integer or name string.
.TP
\f[B]\-\-minlevel\f[R], \f[B]\-L\f[R] \f[I]level\f[R]
Select log messages of the given minimum log level (from 1 to 5).
The argument may be an integer or name string.
.TP
\f[B]\-\-name\f[R], \f[B]\-n\f[R] \f[I]name\f[R]
Name of instance and table in given configuration file (default is
\f[CR]dmfeed\f[R]).
.TP
\f[B]\-\-node\f[R], \f[B]\-N\f[R] \f[I]id\f[R]
Select log messages of the given node id.
.TP
\f[B]\-\-output\f[R], \f[B]\-o\f[R] \f[I]file\f[R]
Path of the output file.
If empty, the Atom feed will be printed to standard output.
.TP
\f[B]\-\-subtitle\f[R], \f[B]\-G\f[R] \f[I]title\f[R]
Sub\-title of feed.
.TP
\f[B]\-\-title\f[R], \f[B]\-C\f[R] \f[I]title\f[R]
Title of feed.
.TP
\f[B]\-\-url\f[R], \f[B]\-U\f[R] \f[I]url\f[R]
Public URL of the feed.
.TP
\f[B]\-\-version\f[R], \f[B]\-v\f[R]
Print version information and quit.
.TP
\f[B]\-\-xsl\f[R], \f[B]\-X\f[R]
URL or path of optional XSLT style sheet.
Web browsers may be able to display the Atom feed in HTML format if a
style sheet is provided.
.SH EXIT STATUS
.TP
\f[B]0\f[R]
Success.
Feed has been created.
.TP
\f[B]1\f[R]
Failure.
Feed creation failed.
.SH EXAMPLE
Generate a unique feed id, and then write the last 50 log messages in
Atom format to file \f[CR]feed.xml\f[R]:
.IP
.EX
$ dmuuid \-p
19c12109\-3e1c\-422c\-ae36\-3ba19281f2e
$ dmfeed \-d /var/dmpack/log.db \-o feed.xml \(rs
  \-I 19c12109\-3e1c\-422c\-ae36\-3ba19281f2e
.EE
.SH SEE ALSO
\f[I]dmuuid(1)\f[R]


CHANGELOG.md
# Change Log

All notable changes to the project will be documented in this file.

## [2.0.0] – Unreleased

## Library

* `Added` error codes
* `Added` _fortran-nng_ interface bindings to NNG to `vendor/`
* `Added` module `dm_arg_parser` (outsourced from `dm_arg`)
* `Added` module `dm_group` for observation groups
* `Added` module `dm_ipc` for socket access
* `Added` module `dm_ipc_async` for asynchronous I/O
* `Added` module `dm_ipc_disco` for service discovery
* `Added` module `dm_ipc_message` for message handling
* `Added` module `dm_ipc_mutex` for mutex access
* `Added` module `dm_ipc_thread` for thread access
* `Added` module `dm_ipc_type` of IPC enumerators
* `Added` module `dm_job_list` (outsourced from `dm_job`)
* `Changed` names of modules `dm_fifo`, `dm_mqueue`, `dm_mqueue_util`, `dm_pipe`,
  `dm_sem`, `dm_signal`, `dm_thread`, `dm_tty` to include `posix` prefix
* `Changed` name of module `dm_system` to `dm_posix`
* `Changed` database schema due to new data models
* `Changed` default database suffix from `.sqlite` to `.db`
* `Changed` GeoCOM API for Lua to accept prototype observation as first argument
* `Changed` model of observation data structure `dm_observ`
* `Deleted` model of request data strucutre `dm_request`

## Programs

* `Changed` response value units in `dmved`

## Documentation

* Converted user guide from AsciiDoc to Markdown
* Converted man pages from AsciiDoc to Markdown

## [1.0.0] – Unreleased

## Library

* `Added` Bessel, Butterworth, Chebyshev filters to `dm_filter`

## [0.9.9] – 2026-01-11

### Library

* `Added` module `dm_coord` for coordinate transformation
* `Added` module `dm_filter` for filtering
* `Added` module `dm_ghostscript` for Ghostscript access
* `Added` module `dm_netstring` for netstring parsing
* `Added` module `dm_statistics` of statistics functions
* `Added` file tree size function to module `dm_file`
* `Added` file modification date/time to function `dm_file_touch()` in `dm_file`
* `Added` format `FORMAT_TSV` for tab-separated values to `dm_format`
* `Added` swap routines to `dm_util`
* `Changed` error code order for future additions
* `Changed` default database cache size of `dm_db_open()` to improve performance

### Programs

* `Added` author option to `dmreport`
* `Added` meta data to PDF output of `dmreport`
* `Fixed` reading of options in `dmdbctl`
* `Changed` last modified date/time to last log date/time in output of `dmfeed`

## [0.9.8] – 2025-08-30

### Library

* `Added` module `dm_image` for image handling
* `Added` module `dm_js` for JavaScript generation
* `Added` module `dm_roff` for GNU roff abstraction
* `Added` module `dm_serial` for derived type serialisation
* `Added` module `dm_transfer` for HTTP-RPC API file transfers
* `Added` GeoJSON Feature Collection procedures to `dm_geojson`
* `Added` image upload to `dm_rpc`
* `Added` HTTP request and response headers to `dm_rpc`
* `Added` HTTP response headers to `dm_fcgi`
* `Added` terminals `gpic`, `postscript`, and `sixeltek` to `dm_plot`
* `Added` image and transfer functions to `dm_db_api`
* `Changed` structure of database abstraction layer, added modules `dm_db_api`, `dm_db_count`, `dm_db_json`, `dm_db_pragma`, and `dm_db_row`
* `Changed` API of modules `dm_arg` and `dm_config` to object-oriented
* `Changed` query parameter API in `dm_cgi`
* `Changed` Git submodules to copies of third-party dependencies
* `Fixed` reading from pipe in `dm_pipe`
* `Fixed` unsigned type conversion in `dm_c`
* `Fixed` return code handling of database select functions

### Programs

* `Added` program `dmcamera` for capturing of camera images
* `Added` program `dmupload` for image upload to HTTP-RPC API
* `Added` scale factor for response values to `dmreport`
* `Added` PDF and PostScript output to `dmreport`
* `Added` message queue reading to `dmmb`
* `Added` image upload to `dmapi`
* `Added` image viewer to `dmweb`
* `Added` environment variables `DM_IMAGE_DB`, `DM_IMAGE_DIR` to `dmapi` and `dmweb`
* `Added` log file output to `dmlogger`
* `Changed` environment variables in `dmapi` and `dmweb` to `DM_BEAT_DB`, `DM_LOG_DB`, `DM_OBSERV_DB`
* `Changed` derived type serialisation in `dmapi` and `dmexport`
* `Changed` database access in `dmexport`
* `Changed` configuration of `dmgrc`

## [0.9.7] – 2025-04-05

### Library

* `Added` module `dm_c` for C interoperability
* `Added` module `dm_camera` for webcam and RTSP access (FFmpeg)
* `Added` module `dm_crypto` of cryptographic hash functions
* `Added` module `dm_db_query` for SQL query building
* `Added` module `dm_dwd` for DWD API format parsing
* `Added` module `dm_dwd_api` for DWD API access (libcurl)
* `Added` module `dm_ftp` for FTP transfer (libcurl)
* `Added` module `dm_freebsd` for system status access on FreeBSD
* `Added` module `dm_geojson` for GeoJSON serialisation
* `Added` module `dm_gm` for GraphicsMagick access
* `Added` module `dm_im` for XMPP connectivity (libstrophe)
* `Added` module `dm_linux` for system status access on Linux
* `Added` module `dm_modbus` for Modbus RTU/TCP connectivity (libmodbus)
* `Added` module `dm_ve` for Victron Energy VE.Direct protocol support
* `Added` library build date and system type parameters (CPP, FPP)
* `Added` Linux and FreeBSD abstractions in module `dm_system`
* `Changed` refactored database abstraction layer
* `Changed` refactored database schema
* `Changed` refactored command-line argument parsing
* `Changed` refactored logical function names
* `Fixed` wrong baud rate in GeoCOM class

### Programs

* `Added` program `dmbot` for remote control of clients through XMPP
* `Added` program `dmdwd` for Deutscher Wetterdienst (DWD) API monitoring
* `Added` program `dmmb` for sensor monitoring through Modbus RTU/TCP
* `Added` program `dmmbctl` for Modbus RTU/TCP command-line control
* `Added` program `dmsystem` for system monitoring
* `Added` program `dmved` for MPPT and battery monitoring (VE.Direct)
* `Added` interactive map view to `dmweb` (Leaflet)

### Documentation


app/dmfeed.f90
! dmfeed.f90
!
! Author:  Philipp Engel
! Licence: ISC
program dmfeed
    !! This program creates a web feed from log messages in Atom
    !! Syndication Format.
    use :: dmpack
    implicit none (type, external)

    character(*), parameter :: APP_NAME  = 'dmfeed'
    integer,      parameter :: APP_MAJOR = 2
    integer,      parameter :: APP_MINOR = 0
    integer,      parameter :: APP_PATCH = 0

    integer, parameter :: APP_MAX_ENTRIES = 500 !! Maximum number of feed entries.

    type :: app_type
        !! Application settings.
        character(ID_LEN)        :: name      = APP_NAME    !! Name of instance/configuration.
        character(FILE_PATH_LEN) :: config    = ' '         !! Path to config file.
        character(FILE_PATH_LEN) :: database  = ' '         !! Path to log database.
        character(FILE_PATH_LEN) :: output    = ' '         !! Output path of Atom file (stdout if empty).
        character(NODE_ID_LEN)   :: node_id   = ' '         !! Optional node id.
        integer                  :: entries   = 50          !! Max. number of entries in feed.
        integer                  :: min_level = LL_DEBUG    !! Minimum log level
        integer                  :: max_level = LL_CRITICAL !! Maximum log level.
        logical                  :: force     = .false.     !! Force writing of output file.
        type(atom_type)          :: atom                    !! Atom type.
    end type app_type

    integer        :: rc  ! Return code.
    type(app_type) :: app ! App settings.

    ! Initialise DMPACK.
    call dm_init()

    ! Read command-line arguments and configuration from file.
    rc = read_args(app)
    if (dm_is_error(rc)) call dm_stop(STOP_FAILURE)

    ! Write Atom XML feed to file or standard output.
    call create_feed(app, rc)
    if (dm_is_error(rc)) call dm_stop(STOP_FAILURE)
contains
    logical function is_stale_file(path, time) result(is)
        !! Returns `.true.` if last modification time of file at `path` is
        !! before `time`.
        character(*),        intent(in) :: path !! Path to file.
        character(TIME_LEN), intent(in) :: time !! ISO 8601 time stamp of last record.

        integer(i8)            :: epoch
        type(file_status_type) :: file_status

        is = .true.

        ! File is stale if it does not exist yet.
        if (.not. dm_file_exists(path)) return

        ! Read file status to get last modification time.
        rc = dm_file_status(app%output, file_status)
        if (dm_is_error(rc)) return
        if (file_status%type /= FILE_TYPE_FILE) return

        ! Convert ISO 8601 to Unix epoch to normalise the GMT offset.
        rc = dm_time_to_epoch(time, epoch)
        if (dm_is_error(rc)) return

        ! Last file modification time is older than given time stamp?
        if (file_status%m_time < epoch) return

        is = .false.
    end function is_stale_file

    subroutine create_feed(app, error)
        !! Creates Atom XML feed from logs in database.
        type(app_type), intent(inout)         :: app   !! App type.
        integer,        intent(out), optional :: error !! Error code.

        integer :: rc
        logical :: is_file

        type(db_type)               :: db
        type(log_type), allocatable :: logs(:)

        is_file = (dm_string_has(app%output) .and. app%output /= '-')

        feed_block: block
            character(TIME_LEN) :: modified
            integer             :: unit, stat

            ! Connect to database.
            rc = dm_db_open(db, app%database, timeout=DB_TIMEOUT_DEFAULT)

            if (dm_is_error(rc)) then
                call dm_error_out(rc, 'failed to open database')
                exit feed_block
            end if

            ! Get logs from database.
            rc = dm_db_select_logs(db        = db, &
                                   logs      = logs, &
                                   node_id   = app%node_id, &
                                   min_level = app%min_level, &
                                   max_level = app%max_level, &
                                   desc      = .true., &
                                   limit     = int(app%entries, i8))

            if (dm_is_error(rc) .and. rc /= E_DB_NO_ROWS) then
                call dm_error_out(rc, 'database error')
                exit feed_block
            end if

            ! Get time stamp of last log record in UTC.
            modified = TIME_DEFAULT
            if (size(logs) > 0) rc = dm_time_to_utc(logs(1)%timestamp, modified)
            app%atom%updated = modified

            if (is_file .and. .not. app%force) then
                ! Write output file only if the time stamp of the last log
                ! record is greater than the file modification time.
                if (.not. is_stale_file(app%output, modified)) then
                    ! Nothing to do here.
                    rc = E_NONE
                    exit feed_block
                end if
            end if

            unit = STDOUT

            if (is_file) then
                rc = E_IO
                open (action='write', file=trim(app%output), iostat=stat, newunit=unit, status='replace')

                if (stat /= 0) then
                    call dm_error_out(rc, 'failed to open file ' // app%output)
                    exit feed_block
                end if

                rc = E_NONE
            end if

            call dm_atom_write(app%atom, logs, unit)
            if (is_file) close (unit)

            ! Set last modification date/time of file to last log date/time.
            if (is_file .and. modified /= TIME_DEFAULT) call dm_file_touch(app%output, modified=modified)
        end block feed_block

        if (present(error)) error = rc
        call dm_db_close(db)
    end subroutine create_feed

    ! **************************************************************************
    ! COMMAND-LINE ARGUMENTS AND CONFIGURATION FILE.
    ! **************************************************************************
    integer function read_args(app) result(rc)
        !! Reads command-line arguments and configuration from file (if
        !! `--config` is passed).
        type(app_type), intent(out) :: app !! App type.

        type(arg_parser_class) :: parser

        call parser%add('name',     short='n', type=ARG_TYPE_ID)      ! -n, --name <id>
        call parser%add('config',   short='c', type=ARG_TYPE_FILE)    ! -c, --config <path>
        call parser%add('database', short='d', type=ARG_TYPE_DATABASE, exist=.true.) ! -d, --database <path>
        call parser%add('output',   short='o', type=ARG_TYPE_FILE)    ! -o, --output <path>
        call parser%add('node',     short='N', type=ARG_TYPE_ID)      ! -N, --node <id>
        call parser%add('entries',  short='E', type=ARG_TYPE_INTEGER) ! -E, --entries <n>
        call parser%add('minlevel', short='L', type=ARG_TYPE_LEVEL)   ! -L, --minlevel <n>
        call parser%add('maxlevel', short='K', type=ARG_TYPE_LEVEL)   ! -K, --maxlevel <n>
        call parser%add('force',    short='F', type=ARG_TYPE_LOGICAL) ! -F, --force
        call parser%add('author',   short='A', type=ARG_TYPE_STRING)  ! -A, --author <string>
        call parser%add('email',    short='M', type=ARG_TYPE_STRING)  ! -M, --email <string>
        call parser%add('id',       short='I', type=ARG_TYPE_STRING)  ! -I, --id <string>
        call parser%add('title',    short='C', type=ARG_TYPE_STRING)  ! -C, --title <string>
        call parser%add('subtitle', short='G', type=ARG_TYPE_STRING)  ! -G, --subtitle <string>
        call parser%add('url',      short='U', type=ARG_TYPE_STRING)  ! -U, --url <string>
        call parser%add('xsl',      short='X', type=ARG_TYPE_STRING)  ! -X, --xsl <string>

        ! Read all command-line arguments.
        rc = parser%read(version_callback)
        if (dm_is_error(rc)) return

        call parser%get('name',   app%name)
        call parser%get('config', app%config)

        ! Read configuration from file.
        rc = read_config(app)
        if (dm_is_error(rc)) return

        ! Get all other arguments.
        call parser%get('database', app%database)
        call parser%get('output',   app%output)
        call parser%get('node',     app%node_id)
        call parser%get('entries',  app%entries)
        call parser%get('minlevel', app%min_level)
        call parser%get('maxlevel', app%max_level)
        call parser%get('force',    app%force)
        call parser%get('author',   app%atom%author)
        call parser%get('email',    app%atom%email)
        call parser%get('id',       app%atom%id)
        call parser%get('title',    app%atom%title)
        call parser%get('subtitle', app%atom%subtitle)
        call parser%get('url',      app%atom%url)
        call parser%get('xsl',      app%atom%xsl)

        ! Validate passed options.
        rc = validate(app)
    end function read_args

    integer function read_config(app) result(rc)
        !! Reads configuration from (Lua) file.
        type(app_type), intent(inout) :: app !! App type.

        type(config_class) :: config

        rc = E_NONE
        if (.not. dm_string_has(app%config)) return

        rc = config%open(app%config, app%name)

        if (dm_is_ok(rc)) then
            call config%get('database', app%database)
            call config%get('output',   app%output)
            call config%get('node',     app%node_id)
            call config%get('entries',  app%entries)
            call config%get('minlevel', app%min_level)
            call config%get('maxlevel', app%max_level)
            call config%get('force',    app%force)
            call config%get('author',   app%atom%author)
            call config%get('email',    app%atom%email)
            call config%get('id',       app%atom%id)
            call config%get('title',    app%atom%title)
            call config%get('subtitle', app%atom%subtitle)
            call config%get('url',      app%atom%url)
            call config%get('xsl',      app%atom%xsl)
        end if

        call config%close()
    end function read_config

    integer function validate(app) result(rc)
        !! Validates options and prints error messages.
        type(app_type), intent(inout) :: app !! App type.

        rc = E_INVALID

        if (dm_string_has(app%node_id) .and. .not. dm_id_is_valid(app%node_id)) then
            call dm_error_out(rc, 'invalid node id')
            return
        end if

        if (.not. dm_string_has(app%database)) then
            call dm_error_out(rc, 'invalid or missing database')
            return
        end if

        if (.not. dm_file_exists(app%database)) then
            call dm_error_out(rc, 'database does not exist')
            return
        end if

        if (.not. dm_log_level_is_valid(app%min_level)) then
            call dm_error_out(rc, 'invalid minimum log level')
            return
        end if

        if (.not. dm_log_level_is_valid(app%max_level)) then
            call dm_error_out(rc, 'invalid maximum log level')
            return
        end if

        if (app%max_level < app%min_level) then
            call dm_error_out(rc, 'maximum level must be greater than minimum level')
            return
        end if

        if (app%entries < 1 .or. app%entries > APP_MAX_ENTRIES) then
            call dm_error_out(rc, 'invalid number of entries')
            return
        end if

        rc = E_NONE
    end function validate

    ! **************************************************************************
    ! CALLBACKS.
    ! **************************************************************************
    subroutine version_callback()
        call dm_version_out(APP_NAME, APP_MAJOR, APP_MINOR, APP_PATCH)
        print '(a, 1x, a)', dm_lua_version(.true.), dm_db_version(.true.)
    end subroutine version_callback
end program dmfeed
